vendor/bat-native-ads/src/bat/ads/internal/ad_events/ad_event_util.h

```diff
@@ -8,12 +8,53 @@
 
 #include "bat/ads/internal/ad_events/ad_event_info.h"
 
+#include <algorithm>
+
+#include "base/time/time.h"
+
 namespace ads {
 
 struct AdInfo;
 
 bool HasFiredAdViewedEvent(const AdInfo& ad, const AdEventList& ad_events);
 
+template <typename T>
+base::Time GetLastSeenAdTime(const AdEventList& ad_events,
+                             const T& creative_ad) {
+  const auto iter = std::find_if(
+      ad_events.begin(), ad_events.end(),
+      [&creative_ad](const AdEventInfo& ad_event) -> bool {
+        return (ad_event.creative_instance_id ==
+                    creative_ad.creative_instance_id &&
+                ad_event.confirmation_type == ConfirmationType::kViewed);
+      });
+
+  double timestamp = 0;
+  if (iter != ad_events.end()) {
+    timestamp = iter->timestamp;
+  }
+
+  return base::Time::FromDoubleT(timestamp);
+}
+
+template <typename T>
+base::Time GetLastSeenAdvertiserTime(const AdEventList& ad_events,
```

Comment:
ditto

Reply:
We should change `const T& creative_ad` to `const CreativeAdInfo& creative_ad` as we will then no longer need the template